File: python/packages/core/src/matimo/tools/matimo_reload_tools/matimo_reload_tools.py
"""matimo_reload_tools — hot-reload tool registry from configured paths.

NOTE: Actual reload is intercepted by Matimo.execute() which calls instance.reload().
This fallback is only reached in environments without the interception.
"""
from __future__ import annotations

import logging

logger = logging.getLogger("matimo")


async def run(params: dict) -> dict:  # noqa: ARG001  # type: ignore[type-arg]
    # Try to call reload on the global instance
    try:
        from matimo.decorators import get_global_matimo_instance

        instance = get_global_matimo_instance()
        if instance is not None:
            await instance.reload()
            tools = instance.list_tools()
            return {
                "success": True,
                "loaded": len(tools),
                "removed": 0,
                "revalidated": 0,
                "rejected": [],
                "message": f"Reloaded {len(tools)} tools successfully.",
            }
    except Exception as exc:
        logger.debug("matimo_reload_tools fallback: %s", exc)

    return {
        "success": False,
        "loaded": 0,
        "removed": 0,
        "revalidated": 0,
        "rejected": [],
        "message": "Reload must be handled by Matimo instance. Initialize Matimo and call matimo.reload() directly.",
    }

File: typescript/packages/core/tools/matimo_reload_tools/matimo_reload_tools.ts
import { getGlobalMatimoLogger } from '@matimo/core';

/**
 * matimo_reload_tools — Hot-reload all tools from configured toolPaths.
 *
 * NOTE: This function is a placeholder. The actual reload logic is intercepted
 * and handled directly by MatimoInstance.execute() because reloadTools() is a
 * method on the instance itself (it clears the registry, re-reads YAML from
 * disk, re-validates untrusted tools, etc.). The function executor cannot do
 * this because it doesn't have a reference to the MatimoInstance.
 *
 * If this file is ever reached (e.g., in tests without the interception),
 * it returns an error instructing the caller to use matimo.reloadTools() directly.
 */
export default async function matimoReloadTools(): Promise<{
  success: boolean;
  message: string;
}> {
  const logger = getGlobalMatimoLogger();
  logger.warn(
    'matimo_reload_tools: reached fallback implementation — this should be intercepted by MatimoInstance.execute()',
  );

  return {
    success: false,
    message:
      'Reload must be handled by MatimoInstance. If you see this, the interception in execute() is not active.',
  };
}
